src/System.IO.FileSystem/src/System/IO/Win32FileStream.cs

```diff
@@ -468,10 +469,13 @@ protected override void Dispose(bool disposing)
                 _canRead = false;
                 _canWrite = false;
                 _canSeek = false;
-                // Don't set the buffer to null, to avoid a NullReferenceException
-                // when users have a race condition in their code (ie, they call
-                // Close when calling another method on Stream like Read).
-                //_buffer = null;
+
+                if (_buffer != null)
+                {
+                    ArrayPool<byte>.Shared.Return(_buffer, clearArray: true);
```

Comment:
@stephentoub that invariant strikes me as an important piece of the security story for managed code -- in general, no matter what (safe, verifiable code) you get wrong, you shouldn't get heap corruption that might be exploitable. 

I don't know enough about this area of the code to say which changes are/aren't affected, but please ensure any areas you're aware of that may cause memory corruption are tracked/fixed.